TrafficSimulator/Assets/RoadGenerator/Core/Runtime/Objects/Road.cs

```diff
@@ -88,59 +123,161 @@ private void UpdateRoad()
                 UpdateRoadNodes();
                 UpdateLanes();
                 roadMeshCreator.UpdateMesh();
+                foreach(Intersection intersection in _intersections)
+                    intersection.UpdateMesh();
                 RoadSystem.UpdateRoadSystemGraph();
                 ShowLanes();
             }
         }
+
+        private void AddIntersectionNode(ref RoadNode curr, Vector3 position, RoadNodeType type)
+        {
+            // TODO: Look into if these distances need to follow the path
+            float distanceToIntersection = Vector3.Distance(curr.Position, position);
+
+            RoadNode next = curr.Next == null ? null : curr.Next.Next;
+
+            float time = next == null ? curr.Time : (next.Time + curr.Time) / 2;
+
+            // Add the intersection node
+            curr.Next = new RoadNode(position, curr.Tangent, curr.Normal, type, curr, next, distanceToIntersection, time);
+            curr = curr.Next;
+
+            // If the road does not end at the intersection, update the distance on the following node as well
+            if(curr != null)
+                curr.DistanceToPrevNode = Vector3.Distance(position, curr.Position);
+        }
         
         /// <summary>Updates the road nodes</summary>
         public void UpdateRoadNodes()
         {
             // Create the vertex path for the road
             BezierPath path = RoadObject.GetComponent<PathCreator>().bezierPath;
             _path = new VertexPath(path, transform, LaneVertexSpacing);
+            
+            this._length = _path.length;
 
             // Set the end of path instruction depending on if the path is closed or not
             this._endOfPathInstruction = path.IsClosed ? EndOfPathInstruction.Loop : EndOfPathInstruction.Stop;
 
             // Create the start node for the road. The start node must be an end node
-            this._start = new RoadNode(_path.GetPoint(0), RoadNodeType.End, 0);
+            this._start = new RoadNode(_path.GetPoint(0), _path.GetTangent(0), _path.GetNormal(0), RoadNodeType.End, 0, _path.times[0]);
             
             // Create a previous and current node that will be used when creating the linked list
             RoadNode prev = null;
             RoadNode curr = _start;
 
             // Calculating the path distance for each intersection on the road
-            List<float> distanceAtIntersection = new List<float>();
+            PriorityQueue<IntersectionVertexPoints> intersectionVertices = new PriorityQueue<IntersectionVertexPoints>();
+            
             foreach(Intersection intersection in _intersections)
             {
-                distanceAtIntersection.Add(_path.GetClosestDistanceAlongPath(intersection.IntersectionPosition));
+                int startIndex = -1;
```

Comment:
This function have become very large and hard to understand. It should be divided into other functions. I also feel like the way you add the intersection nodes and junction nodes are overcomplicated, consider looking at the way i did it in https://github.com/marc7s/TrafficSimulator/commit/67706cc224816288604d4b80d9bff06f12d5f401#diff-b6b29630826789a9e01f0f580805f7a2b0426c5e9435caafffb2e896aa5ba57b, Roads line 134. i believe it should work for three way too.

Reply:
I do not feel any point in doing this. The code does not need improvements to certain functions, it needs to be refactored entirely. It was not written with all scenarios in mind as we were not sure what it would look like, so adding 3-way intersections will always be overly complicated and large as the codebase was not developed with supporting it in mind.

Therefore I do not want to put any time into improvements like these as the intersection code should be completely refactored anyway. However, this code works which means it can be used to continue with the project and develop other areas in the meantime while this is refactored.